buffer/src/main/java/io/netty/buffer/PoolChunk.java

```diff
@@ -643,4 +638,161 @@ static boolean isSubpage(long handle) {
+                int index = index(key);
+                for (int i = 0; i < MAX_PROBE; i++) {
+                    long existing = array[index];
+                    if (existing == key || existing == 0) {
+                        long prev = array[index + 1];
+                        array[index] = key;
+                        array[index + 1] = value;
+                        return prev;
+                    }
+                    index = index + 2 & mask;
+                }
+                expand(); // Grow array and re-hash.
+            }
+        }
+
+        public void remove(long key) {
+            if (key == 0) {
+                zeroVal = 0;
+                return;
+            }
+            int index = index(key);
+            for (int i = 0; i < MAX_PROBE; i++) {
+                long existing = array[index];
+                if (existing == key) {
+                    array[index] = 0;
+                    array[index + 1] = 0;
+                    break;
+                }
+                index = index + 2 & mask;
+            }
+        }
+
+        public long get(long key) {
+            if (key == 0) {
+                return zeroVal;
+            }
+            int index = index(key);
+            for (int i = 0; i < MAX_PROBE; i++) {
+                long existing = array[index];
+                if (existing == key) {
+                    return array[index + 1];
+                }
+                index = index + 2 & mask;
+            }
+            return -1;
+        }
+
+        private int index(long key) {
+            return (int) (key << 1 & mask);
+        }
+
+        private void expand() {
+            long[] prev = array;
+            int newSize = prev.length * 2;
+            mask = newSize - 1;
+            array = new long[newSize];
+            for (int i = 0; i < prev.length >> 1; i += 2) {
+                long key = prev[i];
+                long val = prev[i + 1];
+                put(key, val);
+            }
+        }
+    }
+
+    private static final class LongPriorityQueue {
+        private long[] array = new long[9];
+        private int size;
+
+        public void offer(long handle) {
+            size++;
+            if (size == array.length) {
+                // Grow queue capacity.
+                array = Arrays.copyOf(array, 1 + (array.length - 1) * 2);
+            }
+            array[size] = handle;
+            lift();
+        }
+
+        public void remove(long value) {
+            for (int i = 1; i <= size; i++) {
+                if (array[i] == value) {
+                    if (i == size) {
+                        array[i] = 0;
+                    } else {
+                        array[i] = array[size];
+                        sink(i);
+                    }
+                    size--;
+                    return;
+                }
+            }
+        }
+
+        public long poll() {
+            if (size == 0) {
+                return 0;
```

Comment:
```suggestion
                return -1;
```
I think -1 is better as 0 is a valid value for `offer`

Reply:
If we do this change we also need to adjust a few other places that check for `== 0`